### PR TITLE
Remove reloptions, columnstore, and safefswritesize from pg_appendonly

### PR DESCRIPTION
**Pull request by bmdoil**
This PR removes the reloptions fields from pg_appendonly in favor of writing AO storage options to pg_class.reloptions.
       
If an AO relation has storage, we must record all attributes to pg_class.reloptions even if not specified in the WITH clause since
they are necessary to perform table scans.  
    
In cases where the AO relation does not have storage (partition parent) and an AO storage reloption has been modified
from server defaults, i.e. via the gp_default_storage_options GUC, the parameter is recorded in pg_class.reloptions and used for inheritance purposes only.
     
safefswritesize was used in the past for 'non-mature' filesystems and is currently not used, so remove it.
    
This also removes the following functions that were used to pad AO blocks to filesystem boundaries:
    
AppendOnlyStorageRead_DoSkipPadding
AppendOnlyStorageWrite_DoPadOutRemainder

columnstore field is removed in favor of using the table AM API.

These changes require a catalog version bump.

Discussion: https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/pJ2nqf__xDw

### CODE REVIEW

**Review by huansong — COMMENTED** [review]
Went over the first four commits. Will take a look at the rest today.

**Review by huansong — COMMENTED** [review]
Mostly done. Great work! Good riddance of the other unneeded columns besides the ones in reloptions!

File: src/bin/psql/describe.c [hunk 1]
```diff
@@ -2661,24 +2645,6 @@ describeOneTableDetails(const char *schemaname,
...
-			if (greenplum_is_ao_row(tableinfo.relstorage, tableinfo.relam))
-			{
-				printfPQExpBuffer(&buf, _("Compression Type: %s"), tableinfo.compressionType);
-				printTableAddFooter(&cont, buf.data);
-				printfPQExpBuffer(&buf, _("Compression Level: %s"), tableinfo.compressionLevel);
-				printTableAddFooter(&cont, buf.data);
-				printfPQExpBuffer(&buf, _("Block Size: %s"), tableinfo.blockSize);
-				printTableAddFooter(&cont, buf.data);
-			}
-			printfPQExpBuffer(&buf, _("Checksum: %s"), tableinfo.checksum);
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/psql/describe.c:2678]: I'm thinking that writing checksum to reloptions for `true` or `false` might be the way to go so it behaves the same as the other AO storage options. Thoughts?

File: src/test/regress/expected/create_table_like_gp.out [hunk 2]
```diff
@@ -18,28 +18,16 @@ SET default_table_access_method = ao_row;
...
+select relname, reloptions from pg_class c join pg_am a on a.oid=c.relam where a.amname in ('ao_row','ao_column') AND relname LIKE 't_ao%' order by relname;
+  relname   |                            reloptions                             
+------------+-------------------------------------------------------------------
+ t_ao       | {blocksize=32768,compresslevel=0,compresstype=none}
+ t_ao_a     | {blocksize=32768,checksum=true,compresslevel=0,compresstype=none}
+ t_ao_b     | {blocksize=32768,checksum=true,compresslevel=0,compresstype=none}
+ t_ao_d     | {blocksize=32768,checksum=true,compresslevel=0,compresstype=none}
+ t_ao_enc   | {blocksize=32768,compresslevel=0,compresstype=none}
+ t_ao_enc_a | {blocksize=32768,checksum=true,compresslevel=0,compresstype=none}
+(6 rows)
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/test/regress/expected/create_table_like_gp.out:30]: Good catch, thanks.

File: src/test/regress/expected/dsp.out [hunk 3]
```diff
@@ -331,19 +307,24 @@ ALTER TABLE alter_part_tab1 ADD partition p31 values(1) WITH (appendonly=true, o
...
- alter_part_tab1_1_prt_p2_2_prt_subothers  | t
- alter_part_tab1_1_prt_p31_2_prt_sp1       | t
- alter_part_tab1_1_prt_p31_2_prt_sp2       | t
- alter_part_tab1_1_prt_p31_2_prt_subothers | t
-(9 rows)
+SELECT relname, reloptions from pg_class c where relname like 'alter_part_tab1%';
+                  relname                  |                              reloptions                               
+-------------------------------------------+-----------------------------------------------------------------------
+ alter_part_tab1                           | {checksum=true,compresstype=zlib,blocksize=32768,compresslevel=1}
+ alter_part_tab1_1_prt_p1                  | {blocksize=32768,compresslevel=1,checksum=true,compresstype=zlib}
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/test/regress/expected/dsp.out:314]: Yes, I didn't want to throw away an option if it was explicitly specified, but always writing checksum might make more sense.

File: src/test/regress/expected/alter_table_aocs.out [hunk 4]
```diff
@@ -804,10 +803,10 @@ ALTER TABLE aocs_alter_add_col_no_compress ADD COLUMN d int;
 --------+---------+-----------+----------+---------+---------+--------------+------------------+-------------------+------------+-------------
  a      | integer |           |          |         | plain   |              | none             | 0                 | 32768      | 
  b      | integer |           |          |         | plain   |              | rle_type         | 3                 | 16384      | 
- c      | integer |           |          |         | plain   |              | zlib             | 2                 | 32768      | 
+ c      | integer |           |          |         | plain   |              | none             | 2                 | 32768      |
```
[the remark this answers is not in the document]
Comment (huansong) [reply src/test/regress/expected/alter_table_aocs.out:806]: Oh nvm, since `compresstype` is now always written, so it'll be picked up before the default. So this is the right behavior.

File: src/bin/psql/describe.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (huansong) [reply src/bin/psql/describe.c:2678]: +1 to always write it.

Previously I also thought that saving `checksum=true` is fine but it seems to have created more trouble than whatever benefit it has. Let's just treat it same as others then.

File: src/test/regress/expected/dsp.out [hunk 5]
```diff
@@ -908,17 +832,13 @@ create table ao11 (a int, b int) with (compresstype=none, compresslevel=2)
...
-Block Size: 32768
-Checksum: t
 Distributed by: (a)
 
 set gp_default_storage_options='compresstype=none,compresslevel=2';
 show gp_default_storage_options;
                    gp_default_storage_options                    
 -----------------------------------------------------------------
- blocksize=32768,compresstype=none,compresslevel=2,checksum=true
+ blocksize=32768,compresstype=zlib,compresslevel=2,checksum=true
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/test/regress/expected/dsp.out:841]: `compresslevel=2` is forcing the default `compresstype` here, which is a behavior change. I'll take another look at this but it brings up the question, why do we allow a compresslevel > 0 with `compresstype=none`?

[the remark this answers is not in the document]
Comment (bmdoil) [reply src/test/regress/expected/dsp.out:841]: @huansong 
I reverted the change to force a compression type if `compresslevel` is given but `compresstype=none`.

The issue is with how partition children inherit reloptions from their parents.

Consider
```
 Create table  ao(a int, b text, c int , d int, e numeric,success bool) with ( appendonly=true )
 distributed by (a)
 partition by list(b)
 (
          partition def values ('def','def1','def3') with (appendonly=true, compresslevel=1), 
          default partition dft
 );
```

We expect `def` partition to be `zlib` here, but now that AO reloptions are always written, the parent has `compresstype=none` which is passed down to the child. Since we allow `compresstype=none, compresslevel=1`, the child partition will be created without compression.

Does it still make sense to inherit reloptions from the parent in this scenario?

[the remark this answers is not in the document]
Comment (bmdoil) [reply src/test/regress/expected/dsp.out]: After discussion, we've landed on the following implementation:

If an AO relation has storage, we must record all attributes to pg_class.reloptions even if not specified in the WITH clause since they are necessary to perform table scans.

In cases where the AO relation does not have storage (partition parent) and an AO storage reloption has been modified
from server defaults, i.e. via the gp_default_storage_options GUC, the parameter is recorded in pg_class.reloptions and used for inheritance purposes only.

File: src/bin/psql/describe.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/psql/describe.c:2683]: See https://github.com/greenplum-db/gpdb/pull/14713#discussion_r1080664812

File: src/backend/access/common/reloptions_gp.c [hunk 6]
```diff
@@ -641,54 +643,43 @@ transformAOStdRdOptions(StdRdOptions *opts, Datum withOpts)
 		}
 	}
 
-	if ((opts->blocksize != AO_DEFAULT_BLOCKSIZE) && !foundBlksz)
+
+	/* Record AO storage parameters {blocksize,checksum,compresslevel,compresstype}
```
Comment (huansong) [inline src/backend/access/common/reloptions_gp.c:647]: nit: multi-line comments should start with single-line `/*`

File: src/backend/commands/tablecmds.c [hunk 7]
```diff
@@ -14952,13 +14966,31 @@ ATExecSetRelOptions(Relation rel, List *defList, AlterTableType operation,
...
+			* options over through the AlteredTableInfo struct.
+			*
+			* The exception is partitioned table which won't be rewritten.
+			*/
+			if (*aoopt_changed && rel->rd_rel->relkind != RELKIND_PARTITIONED_TABLE)
+			{
+				ReleaseSysCache(tuple);
+				table_close(pgclass, RowExclusiveLock);
+				return newOptions;
+			}
```
Comment (huansong) [inline src/backend/commands/tablecmds.c:14998]: nit: this seems to be for another commit "Do not update old reloptions during ALTER TABLE in the case of table rewrite"? OK if we are going to squash them.

**Review by huansong — APPROVED** [review]
LGTM besides one CI failure which should be easy to fix. Thanks for addressing all the intricacies! The current solution is very nice.

As discussed, I was thinking about how we can refactor things while reviewing, which can be done later: 

* Make the gp_default_storage_options GUC and the code defaults have the same values. Let them both have `compresstype=none,compresslevel=0,checksum=true,blocksize=32768` to start with.
The only special case is when user changes one of “compresstype” and “compresslevel” in the GUC, we will change the other one in the GUC too, just like how we handle the `WITH` clause.
  * Therefore, the reasoning about what value to pick becomes much simpler: 
    * if something is specified in `WITH`, apply it;
    * apply everything else in the GUC;
    * when `RESET gp_default_storage_options`, we change everything in the GUC back to code defaults.
* Let `default_reloptions` handle the specialties in compression settings as mentioned above.
  * So we can just use `default_reloptions` to process both `WITH` clause and `gp_default_storage_options`. Basically, `gp_default_storage_options` is just like a global `WITH` clause for the same session, which it actually is.
  * But we might need to add an additional parameter to `default_reloptions` to indicate that we are using it for `gp_default_storage_options`, so it won’t pick anything from `gp_default_storage_options` itself.
  * Also with that, `GetAppendOnlyEntryAttributes` just needs to faithfully pass on the values from `default_reloptions`.
* I’m feeling that we might not need to differentiate whether a table `hasStorage` or not now? We might just hide code-default values for normal tables too. But I might be missing something.
* Will revisit some confusing functions. For example `validate_and_adjust_options` and `validateAppendOnlyRelOptions` - are they repetitive and could we reduce/merge them to one?

File: src/backend/commands/tablecmds.c (hunk 7 above)
Comment (bmdoil) [reply src/backend/commands/tablecmds.c:14997]: Moved it to `Do not update old reloptions during ALTER TABLE in the case of table rewrite"`

**Comment by bmdoil** [comment]
Thanks for the review!

> Make the gp_default_storage_options GUC and the code defaults have the same values

Agree, I was surprised when I discovered this wasn't the case.

>Let default_reloptions handle the specialties in compression settings as mentioned above.

I think the separation of upstream and GP reloptions is more trouble than it's worth at this point. Should we consider merging `reloptions_gp.c` into `reloptions.c`? 

> Basically, gp_default_storage_options is just like a global WITH clause for the same session, which it actually is.
On that note, the helper functions to interact with these:
```
/*
 * This is set whenever the GUC gp_default_storage_options is set.
 */
static StdRdOptions ao_storage_opts;
static bool ao_storage_opts_changed = false;
```
should be refactored. In fact ao_storage_opts_changed is always set to true, even if the GUC hasn't been modified. It's also never reset to false, so the bool is functionally useless. I looked into fixing it a bit but decided it was out of scope for this PR.

> I’m feeling that we might not need to differentiate whether a table hasStorage or not now? We might just hide code-default values for normal tables too. But I might be missing something.

The thought was to not hide code-default values for tables with storage in case those code defaults happen to change at some point (i.e. default compression zlib -> zstd).

**Comment by huansong** [comment]
> I think the separation of upstream and GP reloptions is more trouble than it's worth at this point. Should we consider merging reloptions_gp.c into reloptions.c?

Yes the separation in `default_reloptions` could get messy (let alone merging reloptions_gp.c and reloptions.c). But as I think about it again, at least we don't really need an additional parameter in `default_reloptions` just for `gp_default_storage_options`, because we would follow the same rules when picking up options for the GUC and WITH clause: first the passed-in string, then the existing GUC. Then, the compression setting related changes can all be made in `validate_and_refill_options`. So it might be less invasive than I originally thought. But again it's out of the scope of this PR.

> In fact ao_storage_opts_changed is always set to true, even if the GUC hasn't been modified. It's also never reset to false, so the bool is functionally useless. I looked into fixing it a bit but decided it was out of scope for this PR.

Wow I didn't know that. Yes definitely need refactored later. I think after we've made the GUC and code default have the same original value, we will get rid of `ao_storage_opts_changed`. 

> The thought was to not hide code-default values for tables with storage in case those code defaults happen to change at some point (i.e. default compression zlib -> zstd).

Yes I missed that point. Thanks for clarification!
